DigitalLearningSolutions.Data.Tests/Services/UserVerificationServiceTests.cs

```diff
@@ -0,0 +1,340 @@
+
+            // When
+            var (_, verifiedDelegateUsers) = userVerificationService.VerifyUsers(
+                "password",
+                adminUsers,
+                delegateUsers
+            );
+
+            // Then
+            Assert.IsFalse(verifiedDelegateUsers.Contains(secondDelegateUser));
+        }
+
+        [Test]
+        public void VerifyUsers_Returns_no_delegates_if_all_unverified()
+        {
+            // Given
+            A.CallTo(() => cryptoService.VerifyHashedPassword(A<string>._, A<string>._)).Returns(false);
+
+            var firstDelegateUser = UserTestHelper.GetDefaultDelegateUser(1);
+            var secondDelegateUser = UserTestHelper.GetDefaultDelegateUser();
+            var thirdDelegateUser = UserTestHelper.GetDefaultDelegateUser(3);
+
+            var delegateUsers = new List<DelegateUser>
+            {
+                firstDelegateUser,
+                secondDelegateUser,
+                thirdDelegateUser
+            };
+
+            var adminUsers = new List<AdminUser> { UserTestHelper.GetDefaultAdminUser() };
+
+            // When
+            var (_, verifiedDelegateUsers) = userVerificationService.VerifyUsers(
+                "password",
+                adminUsers,
+                delegateUsers
+            );
+
+            // Then
+            Assert.IsEmpty(verifiedDelegateUsers);
+        }
+
+        [Test]
+        public void
+            GetVerifiedAdminUserAssociatedWithDelegateUsers_Returns_nothing_when_delegate_email_is_empty()
+        {
+            // Given
+            var delegateUser = UserTestHelper.GetDefaultDelegateUser(emailAddress: null);
+            var delegateUsers = new List<DelegateUser> { delegateUser };
+
+            // When
+            var returnedAdminUser = userVerificationService.GetVerifiedAdminUserAssociatedWithDelegateUsers(
+                delegateUsers,
+                "password"
+            );
+
+            // Then
+            Assert.IsNull(returnedAdminUser);
+        }
+
+        [Test]
+        public void
+            GetVerifiedAdminUserAssociatedWithDelegateUsers_Returns_nothing_when_no_admin_account_is_associated_with_delegate()
+        {
+            // Given
+            var delegateUser = UserTestHelper.GetDefaultDelegateUser();
+            var delegateUsers = new List<DelegateUser> { delegateUser };
+            A.CallTo(() => userDataService.GetAdminUserByEmailAddress(A<string>._)).Returns(null);
+
+            // When
+            var returnedAdminUser = userVerificationService.GetVerifiedAdminUserAssociatedWithDelegateUsers(
+                delegateUsers,
+                "password"
+            );
+
+            // Then
+            Assert.IsNull(returnedAdminUser);
+        }
+
+        [Test]
+        public void
+            GetVerifiedAdminUserAssociatedWithDelegateUsers_Returns_nothing_when_admin_account_found_is_at_different_centre()

```

Comment:
Think the LoginService should be responsible for this really, and this test should move.